Multiplexed signal implementation

marcelbuesing opened the pull request:
This is currently work in progress. Whats basically missing right now, besides a cleanup, is adding the signal fns to the multiplex structs.

killercup commented on testing/can-messages/src/messages.rs:
```diff
@@ -952,6 +955,179 @@ impl Into<f32> for DolorOneFloat {
+#[derive(Clone, Copy)]
+#[cfg_attr(feature = "debug", derive(Debug))]
+pub enum SensorSonarsSensorSonarsMux<'a> {
+    SensorSonarsSensorSonarsMuxM0(SensorSonarsSensorSonarsMuxM0<'a>),
```
I would be very happy if we could find a shorter/better name for the variants

marcelbuesing replied:
I think for the variants it would probably be sufficient to just name them e.g. M0, M1... for the wrapped structs I think it's probably difficult because of naming collisions.

killercup reviewed (commented):
Some drive-by comments

Like the apporoach of constructing enums with a ref to the raw data slice. I would however like to point out that referencing what is at most 8 bytes with a fat pointer is in the best case (on 32bit ARM) as fast as just copying the whole slice! 😉

marcelbuesing commented on src/lib.rs:
```diff
@@ -7,10 +7,13 @@
 use pad::PadAdapter;
-use std::io::{BufWriter, Write};
+use std::{
+    collections::{HashMap, HashSet},
```
Makes sense I assume it weird when suddenly the generated code looks very different because of small changes in the dbc. I will change this accordingly.

marcelbuesing commented:
> Some drive-by comments
> 
> Like the apporoach of constructing enums with a ref to the raw data slice. I would however like to point out that referencing what is at most 8 bytes with a fat pointer is in the best case (on 32bit ARM) as fast as just copying the whole slice! wink

I think for standard can frames I agree it's probably not worth passing refs compared to copying. It [looks like](https://github.com/marcelbuesing/dbcc/blob/7e3e6b7e608563aa52c1ace143c1194d63b0c183/examples/j1939.dbc#L1511) (example here shows a message with 33 bytes payload) dbc-files is maybe used with CAN-FD (max 64 bytes) or ISO-TP (max 4095 bytes) though. Copy nevertheless as it's maybe an edge case?

killercup commented:
I had assumed that we support just the boring old CAN frames for now, but
you raise a good point. If it's not a big deal and will make this more
future proof, let's go with using references :)

On Sun, 28 Mar 2021, 15:22 Marcel, ***@***.***> wrote:

> Some drive-by comments
>
> Like the apporoach of constructing enums with a ref to the raw data slice.
> I would however like to point out that referencing what is at most 8 bytes
> with a fat pointer is in the best case (on 32bit ARM) as fast as just
> copying the whole slice! wink
>
> I think for standard can frames I agree it's probably not worth passing
> refs compared to copying. It looks like
> <https://github.com/marcelbuesing/dbcc/blob/7e3e6b7e608563aa52c1ace143c1194d63b0c183/examples/j1939.dbc#L1511>
> (example here shows a message with 33 bytes payload) dbc-files is maybe
> used with CAN-FD (max 64 bytes) or ISO-TP (max 4095 bytes) though. Copy
> nevertheless as it's maybe an edge case?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/technocreatives/dbc-codegen/pull/33#issuecomment-808896524>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAE4XYPBPRYX36ML2TRSILTF43SFANCNFSM4Z4FWXKQ>
> .
>

killercup commented on testing/dbc-examples/example.dbc:
```diff
@@ -29,6 +29,17 @@ BO_ 1024 Amet: 8 Sit
 BO_ 1028 Dolor: 8 Sit
  SG_ OneFloat : 0|12@0+ (0.5,0) [0.00|130.00] "" Vector__XXX
 
+BO_ 200 SENSOR_SONARS: 8 SENSOR
```
Is this from a known DBC file/device or did you make it up? I'd prefer to call it something that shows why it's here, like `MULTIPLEX_TEST` with `VARIANT_{A,B,…}_{1,2,…}` signals or something like that

marcelbuesing replied:
I took it from [here](https://github.com/technocreatives/dbc-codegen/issues/19#issuecomment-802653127) but my plan was on changing it before completion to avoid possible copyright issues.

killercup commented on testing/can-messages/src/messages.rs:
```diff
@@ -952,6 +955,522 @@ impl Into<f32> for DolorOneFloat {
+    SensorSonarsSensorSonarsMuxM0(SensorSonarsSensorSonarsMuxM0<'a>),
+    SensorSonarsSensorSonarsMuxM1(SensorSonarsSensorSonarsMuxM1<'a>),
+}
+#[cfg_attr(feature = "debug", derive(Debug))]
```
I know this is just a tiny thing but can you add a newline after the struct/impl blocks?

marcelbuesing replied:
Good observation, just added them

killercup commented on src/lib.rs:
```diff
@@ -215,10 +221,81 @@ fn render_message(mut w: impl Write, msg: &Message, dbc: &DBC) -> Result<()> {
+            match signal.multiplexer_indicator() {
+                MultiplexIndicator::Plain => render_signal(&mut w, signal, dbc, msg)
+                    .with_context(|| format!("write signal impl `{}`", signal.name()))?,
+                MultiplexIndicator::Multiplexor => {
```
please move this block into a function as well before merging :)

killercup commented on src/lib.rs:
```diff
@@ -745,6 +874,72 @@ fn render_debug_impl(mut w: impl Write, msg: &Message) -> Result<()> {
+    let mut multiplexed_signals = HashMap::new();
+    for signal in msg.signals() {
+        if let MultiplexIndicator::MultiplexedSignal(switch_index) = signal.multiplexer_indicator()
+        {
```
Wow I'm surprised rustfmt let you get away with the `{` on the next line

marcelbuesing replied:
Cargo fmt actually moves it to the next line ;), just tried (1.50.0).

killercup replied:
haha, that would make me do `use MultiplexIndicator::*;` :smile:

killercup reviewed (commented):
Hey, sorry for dropping this over easter! How far along is this? I've added some small comments here and there, but I also saw you had a TODO comment. Want to get this into a state where we can merge it even if it's not 100% complete?

marcelbuesing commented:
> Hey, sorry for dropping this over easter! How far along is this? I've added some small comments here and there, but I also saw you had a TODO comment. Want to get this into a state where we can merge it even if it's not 100% complete?

It's usable now, still needs tests against the cantools implementation for verification. I am not particularly happy with the way setting multiplexed signal variants works. See the all.rs `pack_unpack_message_containing_multiplexed_signals` for a demo. It'd be great if anyone has a better idea for this.

killercup commented:
Wow, amazing work!

Fist off: Why did you change all setters to `_raw`? IIRC the raw getter is only useful when you want to raw bits from an enum representation. I'd argue that the default should always be the non-raw version.

Thinking about setters, I'm again conflicted about the borrowing of the data… I think the ideal API is `msg.set_M0(MultiplexTestMultiplexorM0::new(1.0. 1.5)?)`. This does not require creating a signal with a default value and changing it in a closure as it is correct by construction. Do you agree?

We can achieve this by either making the enum own its bytes, or by making it a wrapper around `Cow`. I'd go for making everything more simple and just use the owned data until we run into issues :)

marcelbuesing commented:
> Wow, amazing work!
> 
> Fist off: Why did you change all setters to `_raw`? IIRC the raw getter is only useful when you want to raw bits from an enum representation. I'd argue that the default should always be the non-raw version.
> 
> Thinking about setters, I'm again conflicted about the borrowing of the data… I think the ideal API is `msg.set_M0(MultiplexTestMultiplexorM0::new(1.0. 1.5)?)`. This does not require creating a signal with a default value and changing it in a closure as it is correct by construction. Do you agree?
> 
> We can achieve this by either making the enum own its bytes, or by making it a wrapper around `Cow`. I'd go for making everything more simple and just use the owned data until we run into issues :)

I reverted the _raw suffix changes. I think it would be great in the future to have a more convenient, safer `new` fn that takes e.g. enums as parameters where available and the same for setters. Anyway this should be part of this PR see #39 .

Regardings the setters, I will try it with owned data and see how it turns out, I agree for now it probably should not be an issue and it's better to change it later when the need comes up.

marcelbuesing commented:
Ok I changed it, looks better I think. Any ideas how to do the whole thing without `bitvec::BitArray` for bitwise or ?

killercup commented on testing/can-messages/src/messages.rs:
```diff
@@ -978,6 +982,375 @@ impl Into<f32> for DolorOneFloat {
+    }
+    /// Set value of Multiplexor
+    #[inline(always)]
+    pub fn set_multiplexor(&mut self, value: u8) -> Result<(), CanError> {
```
Aside from the missing newline above the function (:wink:), I'm wondering if we should make this private -- there's little gain by setting just the multiplexor without also adjusting the values, while there's a high chance of producing an invalid state.

marcelbuesing replied:
Yes I actually thought the same ;) it basically asking to be accidentally misused otherwise. Will make this private.

killercup commented on testing/can-messages/src/messages.rs:
```diff
@@ -978,6 +982,375 @@ impl Into<f32> for DolorOneFloat {
+        match self.multiplexor_raw() {
+            0 => MultiplexTestMultiplexor::M0(MultiplexTestMultiplexorM0 { raw: self.raw }),
+            1 => MultiplexTestMultiplexor::M1(MultiplexTestMultiplexorM1 { raw: self.raw }),
+            _ => unreachable!(),
```
This can be reached by user-provided values -- I think we should return a `CanError::InvalidMultiplexor { message_id: 200 }` here.

marcelbuesing replied:
Good point, just fixed it.

killercup reviewed (commented):
Thanks for the quick update again!

killercup commented on src/lib.rs:
```diff
@@ -547,15 +547,19 @@ fn render_multiplexor_signal(mut w: impl Write, signal: &Signal, msg: &Message)
+            writeln!(
+                &mut w,
+                "_ => Err(CanError::InvalidMultiplexor {{ message_id: {}}}),",
+                msg.message_id().0
```
Hmm, looking at this again, do you think we should include the mulitplexor value? Can we always convert it to a u16 or something? Upside: Better debugging. Downside: Larger size of CanError type.

marcelbuesing replied:
Yes I think it's safe to say that it should be possible to convert into u16. I would probably even assume you'll never need more than u8. I think it certainly helps to add the value, otherwise it's hard to debug. So I think it's worth the downside. Went with u16 now, but like i said i can not imagine a use case that requires more than u8 and if there is such a use case that would at least to a compile time error with the generated code.

killercup reviewed (commented):
Thanks for the quick update again!

I saw you fixed some clippy errors, can you also fix the "use of `or_insert` followed by a function call" ones?

killercup commented:
I think this is good to merge besides the small comments above. What do you think?

marcelbuesing commented:
> I think this is good to merge besides the small comments above. What do you think?

Yes I think besides the error type u8/u16 question it should be fine now. I fixed the clippy warnings. Besides that I removed the associated switch index constant as now one can not set the raw value of multiplexor anyway, I don't think it adds value otherwise.

killercup commented:
Alright, let's merge this then!

Thanks again for working on this _for weeks_ -- super impressive!
